// foxify/framework/utils/index.js
"use strict";
function __export(m) {
    for (var p in m) if (!exports.hasOwnProperty(p)) exports[p] = m[p];
}
Object.defineProperty(exports, "__esModule", { value: true });
__export(require("prototyped.js/es6/methods"));
var accepts_1 = require("./accepts");
exports.Accepts = accepts_1.default;
var encode_url_1 = require("./encode-url");
exports.encodeUrl = encode_url_1.default;
var fast_decode_uri_component_1 = require("./fast-decode-uri-component");
exports.decodeURIComponent = fast_decode_uri_component_1.default;
var fresh_1 = require("./fresh");
exports.fresh = fresh_1.default;
var parse_url_1 = require("./parse-url");
exports.parseUrl = parse_url_1.default;
var range_parser_1 = require("./range-parser");
exports.rangeParser = range_parser_1.default;
var vary_1 = require("./vary");
exports.vary = vary_1.default;
function mixins(...baseCtors) {
    return (derivedCtor) => {
        baseCtors.forEach((baseCtor) => {
            // static methods
            Object.getOwnPropertyNames(baseCtor).forEach((name) => {
                if (!exports.array.contains(["length", "constructor", "prototype", "name"], name) && !derivedCtor[name])
                    derivedCtor[name] = baseCtor[name];
            });
            // instance methods
            Object.getOwnPropertyNames(baseCtor.prototype).forEach((name) => {
                if (name !== "constructor" && !derivedCtor.prototype[name])
                    derivedCtor.prototype[name] = baseCtor.prototype[name];
            });
        });
        return derivedCtor;
    };
}
exports.mixins = mixins;
function define(obj, method, name, func) {
    Object.defineProperty(obj, name, {
        configurable: true,
        enumerable: true,
        [method]: func,
    });
}
exports.define = define;
exports.isHandler = (arg) => typeof arg === "function" || arg instanceof Promise;
